SmartPlaces.Facilities/lib/IngestionManager/src/AzureDigitalTwins/AzureDigitalTwinsGraphManager.cs

```diff
@@ -166,43 +167,110 @@ protected virtual async Task ImportRelationshipsAsync(IDictionary<string, BasicR
+            var updateRelationshipSkippedMetricIdentifier = new MetricIdentifier(Metrics.DefaultNamespace, "UpdateRelationshipAsyncSkipped", Metrics.RelationshipTypeDimensionName);
 
             await Parallel.ForEachAsync(relationshipList, parallelOptions, async (relationship, cancellationToken) =>
             {
                 if (queue.TryDequeue(out var digitalTwinsClient))
                 {
                     try
                     {
-                        await digitalTwinsClient.CreateOrReplaceRelationshipAsync(relationship.SourceId, relationship.Id, relationship, cancellationToken: cancellationToken);
-                        TelemetryClient.GetMetric(relationshipsMetricIdentifier).TrackValue(1, relationship.Name);
-                        relationshipCreateSuccesses++;
-                    }
-                    catch (RequestFailedException ex)
-                    {
-                        relationshipCreateFailures++;
+                        BasicRelationship? existingRelationship = null;
+                        try
+                        {
+                            var existingRelationshipResponse = await digitalTwinsClient.GetRelationshipAsync<BasicRelationship>(relationship.SourceId, relationship.Id, cancellationToken);
+                            existingRelationship = existingRelationshipResponse.Value;
+                        }
+                        catch (RequestFailedException ex) when (ex.ErrorCode == "RelationshipNotFound")
+                        {
+                        }
 
-                        // If we are over the limit, sleep for a bit then add the relationship to the retry list
-                        // Todo: jobee - are there any other cases that need to be retried?
-                        if (ex.Status == 429)
+                        if (existingRelationship == null)
                         {
-                            TelemetryClient.GetMetric(throttledMetricIdentifier).TrackValue(1);
-                            Logger.LogError("Connection to ADT Throttled while adding relationship. Sleeping.");
-                            await Task.Delay(options.RetryDelayInMs, cancellationToken);
-                            failedRelationships.TryAdd(relationship.Id, relationship);
+                            try
+                            {
+                                await digitalTwinsClient.CreateOrReplaceRelationshipAsync(relationship.SourceId, relationship.Id, relationship, cancellationToken: cancellationToken);
+                                TelemetryClient.GetMetric(createRelationshipSucceededMetricIdentifier).TrackValue(1, relationship.Name);
+                                relationshipCreateSuccesses++;
+                            }
+                            catch (RequestFailedException ex)
+                            {
+                                relationshipCreateFailures++;
```

Comment:
Should this move inside the !Throttled section? Because I assume we're retrying 429s that until we give up attempting its not a failure to the customer?